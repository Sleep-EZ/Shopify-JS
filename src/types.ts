export const SHOPIFY_TYPE_PRODUCT = 'product';
export const SHOPIFY_TYPE_COLLECTION = 'collection';
export const SHOPIFY_TYPE_PAGE = 'page';

export type ShopifyTypeStr = (
    typeof SHOPIFY_TYPE_PRODUCT |
    typeof SHOPIFY_TYPE_COLLECTION |
    typeof SHOPIFY_TYPE_PAGE);

export type ShopifyType<H extends string> = (
    Product<H> | Collection<H> | Page<H>);

type Expires = { __ts: number };

/**
 * Type definitions for the core Shopify JSON representations.
 *
 * These definitions are based on Shopify's web JSON API,
 * and certainly has the possibility for change (without notice).
 */
export type Product<H> = Expires & {
    id: number,
    handle: H,
    body_html: string | null,
    vendor: string | null,
    product_type: string | null,
    created_at: Date,
    updated_at: Date,
    published_at: Date,
    template_suffix: string,
    published_scope: string,
    tags: string[],
    variants: Variant[],
    options: Option[],
    images: Image[],
    image: Image,
};

export type Image = {
    id: number,
    product_id: number,
    position: number,
    created_at: Date,
    updated_at: Date,
    width: number,
    height: number,
    src: string,
    variant_ids: number[],
};

export type Option = {
    id: number,
    product_id: number,
    name: string,
    position: number,
    values: string[],
};

export type Variant = {
    id: number,
    product_id: number,
    title: string,
    price: number,
    sku: string,
    position: number,
    inventory_policy: string,
    compare_at_price: number,
    fulfillment_service: string,
    inventory_management: string | null,
    option1: string,
    option2: string | null,
    option3: string | null,
    created_at: Date,
    updated_at: Date,
    taxable: boolean,
    barcode: string | null,
    grams: number,
    image_id: number,
    inventory_quantity: number,
    weight: number,
    weight_unit: string,
    inventory_item_id: number,
    old_inventory_quantity: number,
    requires_shipping: boolean,
};

export type Page<H> = Expires & {
    id: number,
    title: string, 
    handle: H,

    body_html: string,
    created_at: Date,
    published_at: Date,
    updated_at: Date,
};

export type Collection<H> = Expires & {
    id: number,
    title: string,
    handle: H,
    description: string,
    published_at: Date,
    updated_at: Date,
    image: Image,
    products_count: number,

    // This is a field provided by this API only
    products: Product<string>[],
}