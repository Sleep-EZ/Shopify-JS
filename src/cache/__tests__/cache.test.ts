import {testCreateCollection, testCreatePage, testCreateProduct} from '../../testing/index';
import {ShopifyTypeEnum} from '../../types';
import {Cache, indexShopifyElement} from '../index';

describe('Cache', () => {
  it('readCache() works', () => {
    const cache = new Cache();
    const testPage = testCreatePage({
      id: 100,
      handle: 'asdas',
    });

    indexShopifyElement(cache._cache, testPage);
    const result = cache.readCache();

    expect(result).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it('can copy data with writeCache()', () => {
    const cache = new Cache();
    const cache2 = new Cache();

    const testPage = testCreatePage({
      id: 100,
      handle: 'asdas',
    });

    indexShopifyElement(cache._cache, testPage);
    cache._cache.ids[100] = null;
    const readCache = cache.readCache();
    expect(readCache).toEqual([null]);

    cache2.writeCache(readCache);
    expect(cache2._cache.ids).toEqual({});
  });

  it('can find an object by ID', () => {
    const cache = new Cache();
    const page = testCreatePage({id: 1});

    indexShopifyElement(cache._cache, page);
    expect(cache._fetchId(1)).toEqual(page);
  });

  it('can handle fetching handles that do not exist', () => {
    const cache = new Cache();
    expect(cache._fetchHandle(ShopifyTypeEnum.Page, 'no-exist')).toEqual(null);
  });

  it('can fetch handles by individual type', () => {
    const cache = new Cache();
    const page = testCreatePage({id: 1, handle: 'example'});
    const product = testCreateProduct({id: 2, handle: 'example'});
    const collection = testCreateCollection({id: 3, handle: 'example'});

    indexShopifyElement(cache._cache, page);
    indexShopifyElement(cache._cache, product);
    indexShopifyElement(cache._cache, collection);

    expect(cache._fetchHandle(ShopifyTypeEnum.Page, 'example')).toEqual(page);
    expect(cache._fetchHandle(ShopifyTypeEnum.Product, 'example'))
        .toEqual(product);
    expect(cache._fetchHandle(ShopifyTypeEnum.Collection, 'example'))
        .toEqual(collection);
  });

  it('can fetch products by handle', () => {
    const cache = new Cache();
    const product = testCreateProduct({
      id: 1,
      handle: 'test',
    });

    cache.set(ShopifyTypeEnum.Product, product);

    expect(cache.getPage('test')).toEqual(null);
    expect(cache.getProduct('test')).toEqual(product);
    expect(cache.getCollection('test')).toEqual(null);
  });

  it('can fetch pages by handle', () => {
    const cache = new Cache();
    const page = testCreatePage({
      id: 1,
      handle: 'test',
    });

    cache.set(ShopifyTypeEnum.Page, page);

    expect(cache.getPage('test')).toEqual(page);
    expect(cache.getProduct('test')).toEqual(null);
    expect(cache.getCollection('test')).toEqual(null);
  });

  it('can fetch collections by handle', () => {
    const cache = new Cache();
    const collection = testCreateCollection({
      id: 1,
      handle: 'test',
    });

    indexShopifyElement(cache._cache, collection);

    expect(cache.getPage('test')).toEqual(null);
    expect(cache.getProduct('test')).toEqual(null);
    expect(cache.getCollection('test')).toEqual(collection);
  });
});
