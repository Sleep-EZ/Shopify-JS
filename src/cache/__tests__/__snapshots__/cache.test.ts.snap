// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cache readCache() works 1`] = `
Array [
  Object {
    "__expires": 1900000000000,
    "__type": "page",
    "body_html": "",
    "created_at": 2010-01-01T00:00:00.000Z,
    "handle": "asdas",
    "id": 100,
    "published_at": 2010-01-01T00:00:00.000Z,
    "title": "",
    "updated_at": 2010-01-01T00:00:00.000Z,
  },
]
`;
